shared/src/clients/phonebook-client.class/index.ts

```diff
@@ -0,0 +1,61 @@
+import axios, { AxiosInstance, AxiosRequestConfig } from 'axios'
+import * as http from 'http'
+
+export default class PhonebookClient {
+  private client: AxiosInstance
+  private baseUrl: string
+  private apiKey: string
+
+  constructor(baseUrl: string, apiKey: string) {
+    this.baseUrl = baseUrl
+    this.apiKey = apiKey
+    this.client = axios.create({
+      baseURL: this.baseUrl,
+      timeout: 5000,
+      headers: {
+        'Content-Type': 'application/json',
+        'x-api-key': this.apiKey,
+      },
+      httpAgent: new http.Agent({ keepAlive: true }),
+    })
+  }
+
+  private request(options: AxiosRequestConfig, body?: any): Promise<any> {
```

Comment:
**[change request]** can we type this instead of using `any`? code smell to use `any` in TS.

Reply:
this is intended as the body data for a request client could be anything. the type is enforced in the function that calls `request`, i.e. the functions below.

Reply:
Hmm so it's not literally anything right, I'm gonna guess at the very least, it'll be a JSON with some compulsory fields (`method`, `url`, etc.)?

Reply:
those mentioned compulsory fields are part of AxiosRequestConfig, the first param, not the body param.

It could be a JSON, it could also be a DTO that we define. Axios handles jsonification for us. 
The body is basically a passthrough param for the `Data`, of type `any`, that is in the axios client, refer to screenshot below.
<img width="548" alt="Screenshot 2023-06-07 at 10 26 33 AM" src="https://github.com/opengovsg/postmangovsg/assets/74128885/608b2a8b-7b20-4b15-84eb-11bad6dc99af">

Reply:
Oops sorry, have misread the code. 

Hmm iinw the `<D = any>` is to provide a default `any` to the generic, not to say that it should be of type `any` though. If we use `any`, we will lose all type-safety 

Seems like the solution here is to use generics. Let me know if any of the below is wrong:
- the second `body` optional argument's type is not known at compile time. this could be the DTO we define.
- the generic that `AxiosRequestConfig` accepts is to provide typing to the data returned

In that case, this should be rewritten as:

```
  private request<TBody, TData>(
    options: AxiosRequestConfig<TData>,
    body?: TBody
  ): Promise<TData> {
    const defaultOptions: AxiosRequestConfig = {
      method: 'post', // default method will be post
    }
    return this.client.request({
      ...defaultOptions,
      ...options,
      data: { ...body },
    })
  }
```

Reply:
on point 2: 
> the generic that AxiosRequestConfig accepts is to provide typing to the data returned
the generic that `AxiosRequestConfig` accepts is to provide typing to the body, not the data.

That said, I am not 100% familiar with generics, but i do understand what you are trying to say. 

I have updated the code to not use `any`. let me know if there are further changes required.

Reply:
Let me look at it again when I reach office.

Reply:
> the generic that AxiosRequestConfig accepts is to provide typing to the body in the request, not the data returned.

Ah you're right. Your change to `options: AxiosRequestConfig<TBody>, body?: TBody` is correct. 

I think there is still a need for the second generic though, using `AxiosResponse` simply hides the fact that `any` is still being returned:
<img width="1009" alt="image" src="https://github.com/opengovsg/postmangovsg/assets/67887489/cb9c90fc-9cc9-4f84-8ede-9815a393334d">

Reply:
updated return type

Reply:
I see you merged alr. It's actually better to use inferred return type (here, to use the second `TData` generic), rather than explicitly add a return type on a per function basis. Right now, your `res.data` is still `any` and you're merely asserting the type in the return (which could be false; [see this](https://www.youtube.com/watch?v=I6V2FkW1ozQ)). 

<img width="660" alt="image" src="https://github.com/opengovsg/postmangovsg/assets/67887489/0b5cc503-22c6-4bbe-9061-42bd47f54b89">

With the second generic and inferred return type, the user of `request` will be required to state the generic, the type will naturally flow through to `res.data`, and this has helped to catch the edge case where you might be sending `data: {}` if `body` is undefined.

<img width="731" alt="image" src="https://github.com/opengovsg/postmangovsg/assets/67887489/ce39e135-da2f-4ba7-b936-f36734be7689">

```TS
request<TData, TBody = any>(
    options: AxiosRequestConfig<TBody>,
    body?: TBody
  ) {
    const defaultOptions: AxiosRequestConfig = {
      method: 'post', // default method will be post
    }
    const requestConfig = body
      ? { ...defaultOptions, ...options, data: { ...body } }
      : { ...defaultOptions, ...options }
    return this.client.request<TData, AxiosResponse<TData>, TBody>(
      requestConfig
    )
  }

  public async getManagedLists(email: string, channel: string) {
    try {
      const res = await this.request<{ id: number; name: string }[]>({
        method: 'get',
        url: `/managed-list`,
        params: {
          owner: email,
          channel,
        },
      })
      return res.data
    } catch (err) {
      throw new Error('Could not get managed lists')
    }
  }

  public async getManagedListById(listId: number) {
    try {
      const res = await this.request<{
        s3Key: string
        etag: string
        filename: string
      }>({
        method: 'get',
        url: `/managed-list/${listId}/members/s3`,
      })
      return res.data
    } catch (err) {
      throw new Error('Could not get managed list by id')
    }
  }
}

```

Reply:
thanks for the in-depth explanation! appreciate that you take the time to write this down and i will be incorporating this into the separate PRs that are coming up as per the above comments.